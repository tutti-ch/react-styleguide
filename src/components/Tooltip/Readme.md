### Tooltip component

The tooltip component is used to display an information for the 
user that might be of help.

```jsx noeditor 
<Tooltip />
```
