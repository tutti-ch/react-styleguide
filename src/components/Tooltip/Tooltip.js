import React, { PureComponent } from "react";
import ToolTipComp from "./_Tooltip";
import Form from "../Form";

export default class Tooltip extends PureComponent {
  reset = e => {
    e.preventDefault();
    localStorage.removeItem("tooltip-styleguide");
    this.tooltip.current.setState({ render: true, closed: false });
  };

  state = {
    selectedArrow: "t"
  };

  tooltip = React.createRef();

  arrowPositions = [
    { text: "Top center", value: "t" },
    { text: "Top right", value: "tr" },
    { text: "Top left", value: "tl" },

    { text: "Bottom center", value: "b" },
    { text: "Bottom right", value: "br" },
    { text: "Bottom left", value: "bl" },

    { text: "Right center", value: "r" },
    { text: "Right top", value: "rt" },
    { text: "Right bottom", value: "rb" },

    { text: "Left center", value: "l" },
    { text: "Left top", value: "lt" },
    { text: "Left bottom", value: "lb" }
  ];

  changeArrow = v => {
    this.setState({ selectedArrow: v });
  };

  render() {
    return (
      <div>
        <div style={{ marginBottom: "1rem" }}>
          <Form.Select
            onChange={this.changeArrow}
            options={this.arrowPositions}
            selected={this.state.selectedArrow}
          />
        </div>
        <ToolTipComp
          name={"tooltip-styleguide"}
          ref={this.tooltip}
          arrowPosition={this.state.selectedArrow}
        >
          This is the tooltip. It will fade away if you click on the close
          button.
        </ToolTipComp>
        <a href={"#"} onClick={this.reset}>
          Click to reset tooltip state
        </a>
      </div>
    );
  }
}
