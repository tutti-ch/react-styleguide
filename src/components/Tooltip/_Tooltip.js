import React, { PureComponent } from "react";
import PropTypes from "prop-types";
import classes from "./_Tooltip.scss";
import classNames from "classnames";
import { Close } from "../../styles/Icons/assets/generic";

export default class Tooltip extends PureComponent {
  static APPEAR_AFTER = 1000; // Will tell the tooltip to appear after this ms

  static defaultProps = {
    arrowPosition: "t",
    appearAfter: Tooltip.APPEAR_AFTER
  };

  static propTypes = {
    /**
     * The position of the arrow.
     */
    arrowPosition: PropTypes.oneOf([
      "t",
      "tr",
      "tl", // top, top-right, top-left
      "b",
      "br",
      "bl", // bottom, bottom-right, bottom-left
      "r",
      "rt",
      "rb", // right, right-top, right-bottom
      "l",
      "lt",
      "lb" // left, left-top, left-bottom
    ]),

    /**
     * The nodes to display in the tooltip.
     */
    children: PropTypes.node,

    /**
     * A unique identifier. If the user has closed the tooltip before,
     * we will not show it again.
     */
    name: PropTypes.string,

    /**
     * Number of ms for the tooltip to appear.
     */
    appearAfter: PropTypes.number
  };

  state = {
    closed: true,
    render: false
  };

  componentDidMount() {
    // Promise is here for unit-testing.
    return new Promise(res => {
      setTimeout(() => {
        if (this.unmounted !== true) {
          this.setState(
            {
              render: !localStorage.getItem(this.props.name), // Render only if the user has not closed before
              closed: false
            },
            res
          );
        }
      }, this.props.appearAfter);
    });
  }

  close = () => {
    // Again, promise is for unit-testing
    return new Promise(res => {
      const { name } = this.props;

      this.setState({ closed: true }, () => {
        if (name) {
          localStorage.setItem(name, "set");
        }

        const time = parseFloat(classes.animationTime) * 1000;
        setTimeout(() => {
          if (this.unmounted !== true) {
            this.setState({ render: false }, res);
          }
        }, time);
      });
    });
  };

  componentWillUnmount() {
    this.unmounted = true;
  }

  render() {
    const { closed, render } = this.state;
    const { arrowPosition, appearAfter, name, children, ...rest } = this.props;

    if (!render) {
      return null;
    }

    const tooltipClasses = classNames(classes.tooltip, classes[arrowPosition], {
      [classes.close]: closed
    });

    return (
      <div {...rest} className={tooltipClasses}>
        {children}
        <div className={classes.closeButton} onClick={this.close}>
          <Close />
        </div>
      </div>
    );
  }
}
