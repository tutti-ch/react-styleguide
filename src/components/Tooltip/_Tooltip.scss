@import "../../styles/main";

$animationTime: 0.25s;

:export {
  animationTime: $animationTime;
}

@keyframes fade-in-sliding {
  0% {
    opacity: 0;
    transform: translateY(8.5rem);
  }
  100% {
    opacity: 1;
    transform: translateY(0);
  }
}

.tooltip {
  background: $blue-50;
  display: inline-block;
  position: absolute;
  padding: $size-s $size-s * 2.5 $size-s $size-s;
  border-radius: $border-radius-m;
  transition: all $animationTime ease-in;
  animation: fade-in-sliding $animationTime ease-in-out;
  opacity: 1;
  color: $greyscale-100;

  &:after {
    position: absolute;
    content: "";
    width: 0;
    height: 0;
  }

  &.t, &.tr, &.tl {
    &:after {
      bottom: 100%;
      border-left: $size-s solid transparent;
      border-right: $size-s solid transparent;
      border-bottom: $size-s solid $blue-50;
    }
  }

  &.b:after,
  &.t:after {
    left: 50%;
    transform: translateX(-50%);
  }

  &.br:after,
  &.tr:after {
    left: 85%;
  }

  &.bl:after,
  &.tl:after {
    right: 85%;
  }

  &.b, &.bl, &.br {
    &:after {
      top: 100%;
      border-left: $size-s solid transparent;
      border-right: $size-s solid transparent;
      border-top: $size-s solid $blue-50;
    }
  }

  &.r, &.rt, &.rb {
    &:after {
      left: 100%;
      border-top: $size-xs solid transparent;
      border-bottom: $size-xs solid transparent;
      border-left: $size-xs solid $blue-50;
    }
  }

  &.l:after,
  &.r:after {
    top: 50%;
    transform: translateY(-50%);
  }

  &.rt:after,
  &.lt:after {
    top: $size-m;
  }

  &.rb:after,
  &.lb:after {
    bottom: $size-m;
  }

  &.l, &.lt, &.lb {
    &:after {
      right: 100%;
      border-top: $size-xs solid transparent;
      border-bottom: $size-xs solid transparent;
      border-right: $size-xs solid $blue-50;
    }
  }

  &.close {
    opacity: 0;
    transform: scale(0);
  }

  .close-button {
    position: absolute;
    right: $size-xxs;
    top: $size-xxs;
    padding: $size-xxs;
    cursor: pointer;

    &:hover :global .svg-sprite {
      fill: $blue-90;
    }
  }

  :global .svg-sprite {
    fill: $greyscale-100;
    width: $size-s;
    height: $size-s;
    margin: 0;
    transition: $transition-m;
  }
}
