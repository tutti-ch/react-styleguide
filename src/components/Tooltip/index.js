export { default } from "./_Tooltip";
