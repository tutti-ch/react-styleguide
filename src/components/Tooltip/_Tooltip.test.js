/* global describe, test, jest, expect, beforeEach */
import React from "react";
import Tooltip from "./_Tooltip";
import { mount } from "enzyme";

describe("(Components) Tooltip", () => {
  test("should first render null then after APPEAR_AFTER should appear and close when clicked on the X", async done => {
    Tooltip.APPEAR_AFTER = 0;
    const tooltipName = "my-tooltip";
    const comp = mount(
      <Tooltip arrowPosition={"tl"} name={tooltipName}>
        My super tooltip
      </Tooltip>
    );
    const inst = comp.instance();
    const spy = jest.spyOn(inst, "close");

    // Initially render is null
    expect(comp.html()).toBe(null);

    // Let's wait till componentDidMount resolves
    await inst.componentDidMount();

    // Update the render
    comp.update();
    expect(comp).toMatchSnapshot();

    // Now let's close the tooltip
    comp
      .find("Close")
      .at(0)
      .closest("div")
      .simulate("click");

    expect(spy).toHaveBeenCalled();

    // Let's wait till close resolves
    await inst.close();
    // Update the render
    comp.update();
    expect(window.localStorage.getItem(tooltipName)).toEqual("set");
    expect(comp).toMatchSnapshot();

    done();
  });

  test("render should be null when user already clicked on close for the tooltip", async done => {
    const tooltipName = "my-tooltip";

    const comp = mount(
      <Tooltip arrowPosition={"tl"} name={tooltipName}>
        My super tooltip
      </Tooltip>
    );
    const inst = comp.instance();

    // Initially render is null
    expect(comp.html()).toBe(null);

    // Let's wait till componentDidMount resolves
    await inst.componentDidMount();

    // Update the render
    comp.update();

    // And should be still null.
    expect(comp.html()).toBe(null);
    done();
  });

  test("on unmount should set a flag so that setState does not get called after unmount", () => {
    const comp = mount(
      <Tooltip arrowPosition={"tl"} name={"lorem ipsum"}>
        My super tooltip
      </Tooltip>
    );
    const inst = comp.instance();
    inst.componentWillUnmount();
    expect(inst.unmounted).toBe(true);
  });
});
