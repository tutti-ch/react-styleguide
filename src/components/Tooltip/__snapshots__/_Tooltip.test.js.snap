// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`(Components) Tooltip should first render null then after APPEAR_AFTER should appear and close when clicked on the X 1`] = `
<Tooltip
  appearAfter={1000}
  arrowPosition="tl"
  name="my-tooltip"
>
  <div
    className=""
  >
    My super tooltip
    <div
      onClick={[Function]}
    >
      <Close>
        <Sprite
          id="close"
          viewBox="0 0 30 30"
        >
          <svg
            className="svg-sprite"
            role="img"
            viewBox="0 0 30 30"
          >
            <use
              xlinkHref="#close"
            />
          </svg>
        </Sprite>
      </Close>
    </div>
  </div>
</Tooltip>
`;

exports[`(Components) Tooltip should first render null then after APPEAR_AFTER should appear and close when clicked on the X 2`] = `
<Tooltip
  appearAfter={1000}
  arrowPosition="tl"
  name="my-tooltip"
/>
`;
